Always show the 404 page

aschempp opened the pull request:
Fixes #4821

we might discuss what features the `PageFinder` should provide.

aschempp committed:
Always show the 404 page

ausi commented on core-bundle/src/Routing/PageFinder.php:
```diff
@@ -0,0 +1,95 @@
+        return $this->framework->getAdapter(PageModel::class)->findPublishedById($pageModel->loadDetails()->rootId);
+    }
+
+    public function findFirstPageTypeByRequestHost(Request $request, string $type): ?PageModel
```
```suggestion
    public function findFirstPageByTypeRequestHost(Request $request, string $type): ?PageModel
```
“find first page type by…” seems confusing to me :)

leofeyer commented:
As discussed in the Contao call, the `PageFinder` should also be used to find the fallback pages in the sitemap controller, the robots.txt controller etc. We do not need to make it `@experimental`.

aschempp committed:
Always use PageFinder to retrieve root pages for request

aschempp commented:
I have updated the implementation. Here are two design-decisions I would like your feedback on @contao/developers 

1. `RobotsTxtController` and `FaviconController` only need the fallback root page for the current host, so they now use `PageFinder` to match the fallback root. `SitemapController` however requires all root pages of the current domain. To be consistent (and after discussing with @Toflar) I decided to go the same way: match the fallback root page for the current domain, and then search the DB for all root pages with the given DNS.

2. The existing controllers previously returned an empty response with status 404 if no root page was found or e.g. the favicon was not configured. I now changed this to throwing a `NotFoundHttpException`, since the whole point of this PR is to render the Contao `error_404` page in such cases.

ausi commented:
> Here are two design-decisions I would like your feedback on

Agree with both decisions 👍

Toflar commented:
Same here, in case that wasn't clear from @aschempp's comment.

ausi commented on core-bundle/src/EventListener/RobotsTxtListener.php:
```diff
@@ -62,14 +62,7 @@ public function __invoke(RobotsTxtEvent $event): void
-        if (null === $rootPage) {
-            return;
-        }
+        $rootPage = $event->getRootPage();
```
Do I understand it correctly, that the deleted code is (and already was) superfluous?

aschempp replied:
yes exactly. I noticed this while reworking the controller.

ausi commented on core-bundle/src/Routing/PageFinder.php:
```diff
@@ -0,0 +1,142 @@
+     *
+     * @return array<PageModel>
+     */
+    public function findAllRootPagesForHost(string $hostname): array
```
```suggestion
    public function findRootPagesForHost(string $hostname): array
```
I think the word “all” is not needed here, and may be confusing because you only get published ones, not all.

Toflar replied:
@aschempp this is unresolved so the review makes no sense.

aschempp replied:
changed in 1e2f600e0a4df54a41b195f4a2ed436df394182f

ausi commented on core-bundle/src/Routing/PageFinder.php:
```diff
@@ -0,0 +1,142 @@
+     */
+    public function findAllRootPagesForRequest(Request $request): array
+    {
+        return $this->findAllRootPagesForHost($request->getHost());
```
```suggestion
    public function findRootPagesForRequest(Request $request): array
    {
        return $this->findRootPagesForHost($request->getHost());
```
Same here

aschempp replied:
changed in 1e2f600e0a4df54a41b195f4a2ed436df394182f

aschempp committed:
Fixed unit tests

aschempp committed:
Fixed unit tests

aschempp committed:
Fixed unit tests

aschempp commented:
I have found a new issue regarding this implementation. Any route that does not have a Contao `frontend` or `backend` scope will not have a firewall (by default). This applies to the image request for example. Not having a firewall means checking for permissions (`Security::isGranted()`) will throw an exception by default in Contao 4. Unfortunately, rendering a 404 page will likely throw such an exception, because a lot of front end modules (e.g. the navigation) will check for user permissions.

There are two possible solution, of which I would prefer the first one:
 1. only implement this change for Contao 5(.2). Would also have the advantage of not adding a new `PageFinder` class in a patch release of Contao 4.13. Since Contao 5 uses Symfony Security 6, a "default firewall" (`NullToken`) is always present.
 2. Patch Symfony security to not throw an exception on Contao 4, basically by faking behaviour of Symfony Security 6 (`NullToken`) but not fully enable the new authentication manager system. Not sure this is really possible or what other side effects that would have on the system 😕

Toflar commented:
I guess introducing this as a new feature is the safer way then. We cannot afford to fiddle around with security in a patch release of the current (and thus most widely adopted and used) LTS version.

aschempp committed:
Use the actual router instead of only the Contao request matchers

aschempp committed:
Added initial tests for PageFinder

aschempp committed:
CI fixes

aschempp committed:
Testing only set input parmameters for main request

aschempp committed:
Fixed tests

aschempp committed:
CI fixes

aschempp committed:
Try to fix the tests

aschempp committed:
Try to fix the CI

Toflar commented on core-bundle/src/Routing/Enhancer/InputEnhancer.php:
```diff
@@ -37,6 +38,12 @@ public function enhance(array $defaults, Request $request): array
             return $defaults;
         }
 
+        $mainRequest = $this->requestStack->getMainRequest();
```
Can you explain this change? What's this doing? (comment in code would make sense then maybe)

aschempp replied:
The goal is to not set input parameters for additional route matching. But I'm not sure this fix is entirely correct, we might need to discuss this problem 😕

aschempp replied:
As discussed at the Contao Call on May 4th, we agreed only the main request should modify Input parameters. I added a comment in 684858bf9e164235821571cd0e0a34c13b024c1e

aschempp committed:
Explain the InputEnhancer changes

aschempp committed:
Merge branch '5.x' into bugfix/symfony-404

aschempp committed:
Simplify the PageFinder methods

leofeyer committed:
Merge branch '5.x' into bugfix/symfony-404

leofeyer committed:
Merge branch '5.x' into bugfix/symfony-404

ausi commented on core-bundle/src/Routing/PageFinder.php:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Finds the first sub-page of a given page for a request host and Accept-Language.
+     * This is mainly useful to retrieve an error page for the current host,
+     * or any other page type that only exists once per root page.
```
```suggestion
     * Finds the first sub-page of a given type for a request host and Accept-Language.
     * This is mainly useful to retrieve an error page for the current host,
     * or any other page type that only exists once per root page.
```
typo?

leofeyer committed:
Fix a typo

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

leofeyer committed:
Merge branch '5.x' into bugfix/symfony-404

leofeyer committed:
CS

leofeyer committed:
Merge branch '5.x' into bugfix/symfony-404

leofeyer commented:
Thank you @aschempp.
